File: 998. Maximum Binary Tree II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
var insertIntoMaxTree = function(root, val) {
    if (root && root.val > val) {         // Since B is A with the value appended to it. Based on the rule, the appended val must be in the left subtree
        root.right = insertIntoMaxTree(root.right, val);
        return root;
    }
    return new TreeNode(val, root);
};

File: Golang/Tree/998.Maximum-binary-tree-ii.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func insertIntoMaxTree(root *TreeNode, val int) *TreeNode {
	if root != nil {
		if root.Val < val {
			return &TreeNode{
				Val:  val,
				Left: root,
			}
		} else if root.Right == nil {
			root.Right = &TreeNode{Val: val}
		} else if root.Right.Val < val {
			root.Right = &TreeNode{
				Val:  val,
				Left: root.Right,
			}
		} else {
			root.Right = insertIntoBST(root.Right, val)
		}
	}
	return root
}
